src/issuance_auth.rs

```diff
@@ -0,0 +1,375 @@
+    type IssueAuthSigType: Clone + PartialEq;
+
+    /// Signs a 32-byte message using the issuance authorizing key.
+    ///
+    /// Only supports signing of messages of length 32 bytes, since we will only be using it
+    /// to sign 32 byte SIGHASH values.
+    fn try_sign(isk: &Self::IskType, msg: &[u8; 32]) -> Result<Self::IssueAuthSigType, Error>;
+
+    /// Verifies a signature over a message using the issuance validating key.
+    fn verify(
+        ik: &Self::IkType,
+        msg: &[u8],
+        signature: &Self::IssueAuthSigType,
+    ) -> Result<(), Error>;
+}
+
+/// An issuance authorizing key.
+///
+/// This is denoted by `isk` as defined in [ZIP 227][issuancekeycomponents].
+///
+/// [issuancekeycomponents]: https://zips.z.cash/zip-0227#issuance-key-derivation
+#[derive(Clone)]
+pub struct IssueAuthKey<S: IssueAuthSigScheme>(S::IskType);
+
+impl<S: IssueAuthSigScheme> IssueAuthKey<S> {
+    /// Sign the provided message using the `IssueAuthKey`.
+    /// Only supports signing of messages of length 32 bytes, since we will only be using it to sign 32 byte SIGHASH values.
+    pub fn try_sign(&self, msg: &[u8; 32]) -> Result<IssueAuthSig<S>, Error> {
+        S::try_sign(&self.0, msg).map(IssueAuthSig)
+    }
+}
+
+/// An issuance validating key which is used to validate issuance authorization signatures.
+///
+/// This is denoted by `ik` and defined in [ZIP 227: Issuance of Zcash Shielded Assets § Issuance Key Generation][IssuanceZSA].
+///
+/// [IssuanceZSA]: https://zips.z.cash/zip-0227#issuance-key-derivation
+#[derive(Clone, PartialEq, Eq)]
+pub struct IssueValidatingKey<S: IssueAuthSigScheme>(S::IkType);
+
+impl<S: IssueAuthSigScheme> IssueValidatingKey<S> {
+    /// Verifies a purported `signature` over `msg` made by this verification key.
+    pub fn verify(&self, msg: &[u8], sig: &IssueAuthSig<S>) -> Result<(), Error> {
+        S::verify(&self.0, msg, &sig.0)
+    }
+}
+
+/// An issuance authorization signature `issueAuthSig`,
+///
+/// as defined in [ZIP 227][issueauthsig].
+///
+/// [issueauthsig]: https://zips.z.cash/zip-0227#issuance-authorization-signature-scheme
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct IssueAuthSig<S: IssueAuthSigScheme>(S::IssueAuthSigType);
+
+/// The Orchard-ZSA issuance authorization signature scheme, based on BIP 340 Schnorr.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ZSASchnorr;
+
+impl IssueAuthSigScheme for ZSASchnorr {
+    const ALGORITHM_BYTE: u8 = 0x00;
+
+    type IskType = NonZeroScalar;
+    type IkType = VerifyingKey;
+    type IssueAuthSigType = schnorr::Signature;
+
+    fn try_sign(isk: &Self::IskType, msg: &[u8; 32]) -> Result<Self::IssueAuthSigType, Error> {
+        schnorr::SigningKey::sign_raw(&schnorr::SigningKey::from(*isk), msg, &[0u8; 32])
+            .map_err(|_| Error::InvalidIssueBundleSig)
+    }
+
+    fn verify(ik: &Self::IkType, msg: &[u8], sig: &Self::IssueAuthSigType) -> Result<(), Error> {
+        ik.verify_prehash(msg, sig)
```

Comment:
We are using `verify_prehash()`, is this the best fit? 
The lib has also ` fn verify_digest(()` and `fn verify()`

If we stick to `verify_prehash`, our API +Docs should reflect it: limit the parameters to the correct size/type

Reply:
I will keep verify_prehash because we will verify a SIGHASH that is already a finalized digest.
verify_digest does not fit because it takes as input a digest that is not finalized
verify does not fit because it hashes the message before verifying it